base/linalg/triangular.jl

```diff
@@ -149,65 +149,98 @@ istril(A::UnitLowerTriangular) = true
 istriu(A::UpperTriangular) = true
 istriu(A::UnitUpperTriangular) = true
 
-function tril(A::UpperTriangular,k::Integer=0)
+function tril!(A::UpperTriangular,k::Integer=0)
     n = size(A,1)
     if abs(k) > n
         throw(ArgumentError("requested diagonal, $k, out of bounds in matrix of size ($n,$n)"))
     elseif k < 0
-        return UpperTriangular(zeros(A.data))
+        fill!(A.data,0)
+        return A
     elseif k == 0
-        return UpperTriangular(diagm(diag(A)))
+        for j in 1:n, i in 1:j-1
```

Comment:
I think it's all correct and good to merge now. Making it more concise can be left for later. Someone also needs to fix the sub-vs-superdiagonal thing in the docs but that doesn't have to be done here. I'll do it if no one beats me to it and I don't forget about it.